### ISSUE TITLE
Glitches with the Prawnsuit, Seamoth, and Cyclops 

### ISSUE THREAD

**Issue by ItzKyros** [SubnauticaNitrox/Nitrox#1359]
### Version

 1.4.0.0
### _Seamoth:_

**1. Random Explosions**

- At random times the Seamoth will explode for no reason.
- This can happen when someone is in the vehicle or while they aren't present.

**2. Turning Invisible**

- This seems to also happen at random times however it only occurs for one player.
- There are some instances where someone enters the Seamoth and both the Seamoth and the player become invisible to other players.
- After this glitch occurs, my friend and I leave the game and join back which normally fixes the issue. However, when joining back, one of us or both of us spawn at a random location on the map. This can range from outside the map to 100 meters away from where the player left.
- Once the player leaves the vehicle, both the Seamoth and player become visible.

**3. Modules only appear and work for one player**

- Putting an upgrade module in the Seamoth will only work for the player who did it. Anyone else who tries to use the Seamoth won't see or be able to use the upgrades.

**4. Entering the vehicle facing sideways**

- Sometimes when you enter the Seamoth, you face a random direction at an awkward angle. 
- In order to fix it you have to leave the vehicle and re-enter.

**5. Seamoth gets stuck in place**

- Sometimes when entering a Seamoth and trying to drive forward the vehicle gets stuck in place. This only occurs for one player.
- In order to fix this glitch someone else has to get in and drive the Seamoth and then get back out. Then the other player can move the Seamoth again.

### _Prawnsuit:_

**1. Upgrades only appear and work for 1 player at a time.**

- This is the same bug listed in number 3 for the Seamoth.

**2. Does not charge**

 - When loaded into Cyclops, the Prawnsuit will continue to take power from the Cyclops trying to charge itself. Power is drained from the Cyclops but the Prawnsuit never receives the power.
- Essentially, the Prawnsuit constantly draws power from the Cyclops to charge itself but never actually charges. This results in the Cyclops losing its power for no reason at a rate of about 1% per second.

**3. Frozen player in the moonpool**

- Players may randomly freeze when they enter the Prawnsuit in the moonpool
- When a player enters the Prawnsuit while it's parked in the moonpool, sometimes they appear frozen on another player's screen. For them, it appears that they are moving around in the Prawnsuit doing their own business. However, for another player, the person in the Prawnsuit appears frozen hanging inside the moonpool. 
- When this happens, the frozen player seems to be greyed/blacked out,
- The fix for this seems to be for the players to leave and rejoin the game.
- When they join back, sometimes it teleports them to a random part of the map.
- There are also some instances during this bug in which a player will instantly die upon entering the water below the moonpool.

**4. Drill arm animation and large resource deposits.**

- When a player uses the prawn suit drill arm, the drilling animation is only visible for the person in the prawn suit.
- In addition to this, using the drill on the deposit does not show the deposit being broken for other players. For other players, the deposit is still visible and appears as if nothing happened to it.

### _Cyclops:_

**1. Cyclops shakes for the players who aren't driving**

- There is a lot of shaking for other passengers which can make it a bit annoying.
- For the driver, the other players appear with a running animation when they shake.

**2. Players randomly launched out of the vehicle**

- This glitch mostly happens when another player starts to drive.
- The player is either: Launched thousands of meters away, teleported under the vehicle, or instantly killed.

**3. Random occurrences where the vehicle appears to flip over on a player's screen.**

- This glitch seems to only occur for one of the players. It can be fixed by leaving and rejoining the game.

**4. Seamoth can be forced into the Cyclops' vehicle storage which breaks the game physics**

- Sometimes the Seamoth can be rammed into the Cyclops so that it glitches into it. When this happens all air acts as water which lets the player swim inside the Cyclops as if it was filled with water. 
- In order to fix this the player must re-enter the Seamoth and then exit it.

### Steps to reproduce

Most of these bugs randomly take place so there's no easy way to test them out. The easiest way would be to spawn the vehicles in a multiplayer game with 2 people and experiment with the bugs mentioned above.

**Comment by AntalaFilip** [comment]
Can confirm all of the Seamoth points, and all of the Prawn except point 2 and
> There are also some instances during this bug in which a player will instantly die upon entering the water below the moonpool.
And as far as I know, the first two Cyclops points are just there for now, you should place a chair for the other players to sit in.

**Comment by neutrinos4all** [comment]
I can also confirm all of the bugs mentioned for the Seamoth and the Cyclops, and in fact I had logged in to report a selection of them after testing but found this issue instead.

**Comment by gir489** [comment]
I confirmed all these behaviours in the recent build. All erroneous. The Prawn upgrade I found can be fixed by pulling out the upgrades and putting them back. Sometimes the Cyclops does the same thing. The prawn power consumption from the cyclops can be fixed by setting the battery to exactly 100% before entering it in to the cyclops, but that requires a debugger to accomplish.

**Comment by ItzKyros** [comment]
The last glitch seems to have only happened twice throughout the entire playthrough and I'm going to give a few more details on it. Basically, when you're loading your Seamoth into the Cyclops and there isn't much room (for example one of the times this glitch occurred was when the Cyclops was parked very low to the ground which didn't give the Seamoth much clearance), the Seamoth might try to force itself into the docking bay by traveling through the wall. After this, when you enter the Cyclops, you are able to swim inside of it even though it's completely filled with air. If you go back to the Seamoth and re-enter it and then go back into the Cyclops, the glitch fixes itself.

**Comment by AntalaFilip** [comment]
> I confirmed all these behaviours in the recent build. All erroneous. The Prawn upgrade I found can be fixed by pulling out the upgrades and putting them back. Sometimes the Cyclops does the same thing. The prawn power consumption from the cyclops can be fixed by setting the battery to exactly 100% before entering it in to the cyclops, but that requires a debugger to accomplish.'
> 
> The cyclops jittering problem is from you not building them a chair.
> 
> Here are some video clips:
> 
> https://youtu.be/sKMxVvsyowE
> https://www.youtube.com/watch?v=-D-MSbHUlYw

It seems that the vehicle upgrades only sync when the players are online - if a player is not online when the upgrade gets installed it doesn't sync.

**Comment by mpocius** [comment]
It's relatively minor compared to the major issues listed here, but I also noticed that if one player in a seamoth gets a crush depth warning, the warning will be played for a second player in a seamoth as well.

**Comment by AntalaFilip** [comment]
> It's relatively minor compared to the major issues listed here, but I also noticed that if one player in a seamoth gets a crush depth warning, the warning will be played for a second player in a seamoth as well. 

Yup, I think that's due to the sound being synchronized between all players.

**Comment by Exso12Ecesis** [comment]
If you have the Seamoth container upgrade and store items in the container, the items will be lost upon server restart.

**Comment by Exso12Ecesis** [comment]
While investigating #1174, I found that the cause of the issue is that in the EntityData.JSON, the container for each seamoth container for the client desyncs when another user inserts an upgrade. Both clients temporarily see both upgrades for this session, but upon restart, only the client who inserted them still has the upgrades in the seamoth. The fix should be to sync both client's container data when another user inserts an upgrade. This may not be possible on far away targets...

**Comment by Exso12Ecesis** [comment]
Another bug you do not have documented, that I have encountered while testing, is that in any of the alien structures, the Prawn constantly gets stuck. If you do not have another client near you, or the grappling hook arm attachment, the Prawn is basically FUBAR and can not be retrieved. Commands like warp and tp have no effect.

**Comment by AntalaFilip** [comment]
> 3. Frozen player in the moonpool
This also happens with the Seamoth, not just the Prawn

**Comment by romgerman** [comment]
If you dock seamoth or prawnsuit (with 100% battery charge) into cyclop it will drain all cyclopes power (like 1% per second).

**Comment by Exso12Ecesis** [comment]
> If you dock seamoth or prawnsuit (with 100% battery charge or less) into cyclop it will drain all cyclopes power (like 1% per second).

That is not true, you have to set it to EXACTLY 100%, there is a check inside the power recharging function in Subnautica (not Nitrox) that checks if the power is == 100.f. The moment you take the vehicle out of the docking station, it begins consuming power based on the number of cycles (ticks/frames/whatever the game uses) to drain power equally. You will notice if you attach a debugger to Subnautica, you will see the power is actually 99.something but actually displays 100% to the user. This is just rounding at work. The end-user does not care it is actually a fraction of a percentage. To fix this issue, attach a debugger to Subnautica, and find the power percentage by using memory manipulation techniques. Once you have found the address, just freeze it to 100.f and then dock it. The power will no longer be consumed from the Cyclops.

I know I just wrote a diatribe that NOBODY but me understood, so in the essence of the internet.

TL;DR: https://fearlessrevolution.com/viewtopic.php?t=5612 Download this, run Inf. Battery, stick the prawn in, bada bing bada boom solved.

**Comment by romgerman** [comment]
@Exso12Ecesis so that's a bug inside Subnautica? Well, the good thing is that the mod devs could potentially fix it inside the mod.

**Comment by Exso12Ecesis** [comment]
> @Exso12Ecesis so that's a bug inside Subnautica? Well, the good thing is that the mod devs could potentially fix it inside the mod.

The team just haven't gotten around to it. Look at how much crap is broken with Nitrox, and it's very complicated. Trust me, they're working on it. But for now, the solution I posted works to keep on truckin'.

**Comment by Measurity** [comment]
Closing this because it has multiple issues in one issue which make it hard to keep track of when it's fixed.
